### DISCUSSION TITLE
Can not connect to cockroachdb cloud

### DISCUSSION THREAD

**Discussion by iamchathu** [mikro-orm/mikro-orm#2172] [Q&A] [answered]
**Describe the bug**
Can not connect to CockrochDB cloud instance. (100% Postgresql compatible database.)

I checked with GUI client and the the options I have specified are correct.

**Stack trace**
```bash
[discovery] ORM entity discovery started, using ReflectMetadataProvider
[discovery] - processing 1 files
[discovery] - processing entity Category (/Users/Chathu/Workspace/GeniusSoft/SmartPOS/smartpos-sever/src/entities/category.ts)
[discovery] - entity discovery finished, found 1 entities, took 207 ms
[info] MikroORM failed to connect to database <clustername>.defaultdb on postgresql://<username>:*****@free-tier.gcp.cockroachlabs.cloud:26257
```

When run migration with same config details.

```bash
error: server requires encryption
    at Parser.parseErrorMessage (/Users/Workspace/project/node_modules/pg-protocol/src/parser.ts:369:69)
    at Parser.handlePacket (/Users/Workspace/project/node_modules/pg-protocol/src/parser.ts:188:21)
    at Parser.parse (/Users/Workspace/project/node_modules/pg-protocol/src/parser.ts:103:30)
    at Socket.<anonymous> (/Users/Workspace/project/node_modules/pg-protocol/src/index.ts:7:48)
    at Socket.emit (events.js:315:20)
    at Socket.EventEmitter.emit (domain.js:467:12)
    at addChunk (internal/streams/readable.js:309:12)
    at readableAddChunk (internal/streams/readable.js:284:9)
    at Socket.Readable.push (internal/streams/readable.js:223:10)
    at TCP.onStreamRead (internal/stream_base_commons.js:188:23) {
  length: 47,
  severity: 'FATAL',
  code: '08004',
  detail: undefined,
  hint: undefined,
  position: undefined,
  internalPosition: undefined,
  internalQuery: undefined,
  where: undefined,
  schema: undefined,
  table: undefined,
  column: undefined,
  dataType: undefined,
  constraint: undefined,
  file: undefined,
  line: undefined,
  routine: undefined
}
```

**To Reproduce**
1. Create sample project.
2. Create free account at Cockroach.
3. Start a free server.

**Expected behavior**
To connect with DB without issues.

**Additional context**

my config

```ts
import { MikroORM } from '@mikro-orm/core';
import { PostgreSqlDriver } from '@mikro-orm/postgresql';
import * as fs from 'fs';
import config from '.';

const getOrm = async (): Promise<MikroORM<PostgreSqlDriver>> =>
  MikroORM.init({
    driver: PostgreSqlDriver,
    driverOptions: {
      ssl: {
        cert: fs.readFileSync(`certs/cc-ca.crt`).toString(),
        rejectUnauthorized: false,
      },
    },
    entities: ['./dist/entities/**/*.js'],
    entitiesTs: ['./src/entities/a**/*.ts'],
    type: 'postgresql',
    dbName: config.database.database,
    name: config.database.database,
    host: config.database.host,
    port: config.database.port,
    user: config.database.user,
    password: config.database.password,
    timezone: '+05:30',
    debug: process.env.NODE_ENV !== 'production',
  });

export default getOrm;

```

**Versions**

| Dependency | Version |
| - | - |
| node | v14.15.5 |
| typescript | ^4.4.2 |
| mikro-orm | ^4.5.9 |
| @mikro-orm/postgresql | ^4.5.9 |

**Comment by B4nan** [comment, marked as answer]
You have the driverOptions wrong, it is the top level setting for knex, you are missing the connection object there. 

See for example https://github.com/mikro-orm/mikro-orm/issues/303 (there are more similar issues if you search for pg and ssl).

**Reply by iamchathu** [reply]
@B4nan Thanks for the reply. 

I got 

```bash

/Users/project/node_modules/knex/lib/dialects/postgres/index.js:148
        resolver(/^PostgreSQL (.*?)( |$)/.exec(resp.rows[0].version)[1]);
                                                                    ^
TypeError: Cannot read property '1' of null
```

**Reply by iamchathu** [reply]
Seems I also need to set `version: '7.2',` also in driver options

**Reply by kenny007** [reply]
@iamchathu please how did you resolve this? I am also having the same issue. This is my setting
`import { Options } from '@mikro-orm/core'
import { PostgreSqlDriver } from '@mikro-orm/postgresql'
import dotenv from 'dotenv'
import fs from 'fs'

dotenv.config()

function getSSLConfig() {
  const caPath = process.env.COCKROACHDB_CA_CERT_FILE
  console.log(`Using CA certificate file: ${caPath}`)
  const sslConfig: any = {
    rejectUnauthorized: Boolean(
      process.env.COCKROACHDB_CA_CERT_FILE ||
        process.env.COCKROACHDB_CA_CERT_DATA,
    ),
  }

  if (process.env.COCKROACHDB_CA_CERT_FILE) {
    sslConfig.ca = fs
      .readFileSync(process.env.COCKROACHDB_CA_CERT_FILE)
      .toString()
  } else if (process.env.COCKROACHDB_CA_CERT_DATA) {
    sslConfig.ca = process.env.COCKROACHDB_CA_CERT_DATA
  }

  return sslConfig
}

const sslFromConfig = getSSLConfig()

console.log(sslFromConfig)

const config: Options = {
  entities: ['./modules/**/**/*.entity.ts'],
  dbName: process.env.DB_NAME || 'dev',
  driver: PostgreSqlDriver,
  migrations: {
    path: './migrations',
    tableName: 'mikroorm_migrations',
    transactional: true,
    emit: 'ts',
  },
  clientUrl: process.env.COCKROACHDB_CONNECTION_STRING,
  driverOptions: {
    connection: {
      client: 'pg',
      version: '13.0.0',
      ssl: {
        ca: sslFromConfig.ca,
      },
    },
  },
  metadataCache: {
    enabled: true,
    pretty: false,
  },
  debug: process.env.NODE_ENV !== 'production',
  allowGlobalContext: true,
  pool: { min: 1, max: 10 },
}

export default config`

**Reply by kenny007** [reply]
This is the error message
`return /^PostgreSQL (.*?)( |$)/.exec(versionString)[1];`

**Reply by kenny007** [reply]
@iamchathu @B4nan  I am having a similar issue currently. is there a way to resolve it? This is my config options.
```suggestion const config: Options = {
  entities: ['./modules/**/**/*.entity.ts'],
  dbName: process.env.DB_NAME || 'dev',
  driver: PostgreSqlDriver,
  migrations: {
    path: './migrations',
    tableName: 'mikroorm_migrations',
    transactional: true,
    emit: 'ts',
  },
  clientUrl: process.env.COCKROACHDB_CONNECTION_STRING,
  driverOptions: {
    connection: {
      client: 'pg',
      version: '13.0.0',
      ssl: {
        ca: sslFromConfig.ca,
      },
    },
  },
  metadataCache: {
    enabled: true,
    pretty: false,
  },
  debug: process.env.NODE_ENV !== 'production',
  allowGlobalContext: true,
  pool: { min: 1, max: 10 },
}

**Reply by giancarlo-dm** [reply]
@kenny007 or whoever faces this issuse the correct `driverOptions` needs to be.
```
driverOptions: {
	client: 'cockroachdb', // this is not inside connection!
	connection: {
		ssl: true
	}
}
```

@B4nan please add this to the docs in a way that we can easily search by "cockroach" and make a quick reference [here ](https://mikro-orm.io/docs/configuration#driver) when it says "compatible with CockroachDB".

**Reply by kenny007** [reply]
@giancarlo-dm I tried this previously and it didn't work.

**Comment by duckies** [comment]
I recently have made a similar mistake of putting mysql connection options at the top level of `driverOptions` because of my interpretation of this fragment of the installation docs:

> We can pass additional options to the underlying driver (e.g. `mysql2`) via `driverOptions`. The object will be deeply merged, overriding all internally used options.

Perhaps the documentation could clarify that `driverOptions` are actually options for knex, with `drierOptions.connection` containing the underlying driver configuraton?

**Reply by B4nan** [reply]
PR welcome. I am also not very happy with the naming, as "driver" is a term used by the ORM too, I am usually referring the pg/mysql2/sqlite3 libraries as "connectors". Note that `driverOptions` are also available for mongo, there is no knex involved, so the new copy should probably mention that too.

**Comment by B4nan** [comment]
#7171 adds a guide about cockroachdb (v7)
